hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/cache/FullTableCache.java

```diff
@@ -111,18 +116,31 @@ public void loadInitial(CacheKey<KEY> key, CacheValue<VALUE> value) {
   @Override
   public void put(CacheKey<KEY> cacheKey, CacheValue<VALUE> value) {
     try {
-      lock.writeLock().lock();
+      lock.readLock().lock();
       cache.put(cacheKey, value);
-      epochEntries.computeIfAbsent(value.getEpoch(),
-          v -> new CopyOnWriteArraySet<>()).add(cacheKey);
+      // add in case of null value for cleanup purpose only when key is deleted
+      if (value.getCacheValue() == null) {
+        epochEntries.computeIfAbsent(value.getEpoch(),
+            v -> new CopyOnWriteArraySet<>()).add(cacheKey);
+      }
     } finally {
-      lock.writeLock().unlock();
+      lock.readLock().unlock();
     }
   }
 
   @Override
   public void cleanup(List<Long> epochs) {
-    executorService.execute(() -> evictCache(epochs));
+    epochCleanupQueue.clear();
```

Comment:
What will happen to the `epochs` that were there in the `epochCleanupQueue` which were not picked up by the `cleanupTasks`?

Reply:
We are giving preference to latest epoch, and since latest epoch is received, which means previous epoch is already notified. Using this, intermediate epoch is not required and in eviction logic, only latest is used to handle all previous epch and action for cleanup.
Check for sequential epoch is not required and logic is modified.